### PR TITLE
Katie and Sabine's Ride Share!

### PR DESCRIPTION
**Pull request by katietj**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | When we were making the new driver class that inherited from user, we considered using Super in the initialize method. We did not end up using that during out initial build of the method, but while refactoring the method, we found that using super made the code DRY-er. 
Describe any examples of composition that you encountered in this project, if any | In trip_dispatcher.rb, the methods find_driver and find_passenger are very similar methods and don't use inheritance. 
Describe the relationship between `User` and `Driver`| The ' Driver' class inherited  data from the 'User' class in this project. The Driver is a child/subclass while the the User is parent/superclass. 
Describe a _nominal test_ that you wrote for this assignment. | In driver_spec.rb, line 148, we created a nominal test by confirming that total net expenditure is returned.
Describe an _edge case test_ that you wrote for this assignment | In trip_dispatch.spec.rb, line 137 , we created a edge case were we eliminated all : AVAILABLE drivers and set the test to confirm that the method that method would raise and ArgumentError.
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | This project enhanced our collective understanding of the relationships between classes, modules, instances and inheritance overall. It also provided us with additional skills around TDD, particularly in the phases related to the creation of tests and code writing ... as well as refactoring.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We both agreed that taking the time to fully understand the assignment, create UMLs and pseudocodes prior to writing codes would have provided more background and framework. For instance focusing on the 'visual' design and parsing out the multiple spec, test,  and csv files would have set us up for success and allowed us to both understand and participate to our fullest potential and ability.

### CODE REVIEW

File: specs/trip_dispatch_spec.rb [hunk 1]
```diff
@@ -111,4 +114,34 @@
...
+      expect(trips_plus_one).must_equal original_trips + 1
+
+    end
+
+    it "should throw argument error if no driver is avalible" do
+      @dispatcher.find_driver(5).driver_on_trip
+      @dispatcher.find_driver(8).driver_on_trip
+      expect{@dispatcher.request_trip}.must_raise ArgumentError
+    end
+    it "should throw argument error if a driver is trying to drive themselves" do
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:142]: You should also have a test where the only driver available requests a trip

File: specs/user_spec.rb [hunk 2]
```diff
@@ -34,15 +34,60 @@
...
+      start_time: Time.parse("2016-08-08"),
+      end_time: Time.parse("2016-08-09"),
+      cost: 10, rating: 5)
+      @user.add_trip(trip_1)
+      @user.add_trip(trip_2)
+      @user.add_trip(trip_3)
+      expect(@user.net_expenditures).must_equal (10 + 10 + 10)
+    end
+
+    it "calculates total trip time" do
```
Comment (CheezItMan) [inline specs/user_spec.rb:61]: You should also test the total duration when there are no trips.

File: specs/driver_spec.rb [hunk 3]
```diff
@@ -77,28 +83,70 @@
...
+      trip2 = RideShare::Trip.new(id: 8, driver: "other_driver", passenger: nil,
+      start_time: Time.parse("2016-08-08"),
+      end_time: Time.parse("2016-08-09"),cost: 5,
+      rating: 1)
+
+      @driver.add_driven_trip(trip1)
+      @driver.add_trip(trip2)
+    end
+
+    it "returns total correct net expenditure" do
```
Comment (CheezItMan) [inline specs/driver_spec.rb:148]: What about when there are no trips, or no driven trips?

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Good number of commits, very good commit messages.  Only one partner made the commits.
Answer comprehension questions | Check, well done
**Wave 1** |
Appropriate use of Ruby's `Time`| Check
`Trip` has a helper method to calculate duration| Check
`User` (passenger) has a method to calculate total cost of all trips| Check
Tests for wave 1 | Check
**Wave 2** |
`Driver` inherits from `User` | Check
`Driver` has `add_driven_trip` method | Check
`Driver` has method to calculate average rating | Check
`Driver` has method to calculate net expenditures and it uses `super` | Check
`Driver` has a method to calculate total revenue | Check
Tests for wave 2 | Check
**Wave 3** |
`TripDispatcher` has a new method to create trips | Check
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | Check
Complex logic was correctly implemented | Check
Tests for `request_trip` | Check, overall well done
Methods from wave 1 and 2 handle incomplete trips | Check 
Tests for wave 1 and 2 methods with incomplete trips | Missing
**Overall** | Overall pretty well done, see my inline notes.  You hit all the learning goals for the project.
